pallets/oracle/src/dia.rs

```diff
@@ -0,0 +1,121 @@
+const KSM_DIA_SYMBOL: &str = "KSM";
+pub struct DiaOracleKeyConvertor;
+impl Convert<OracleKey, Option<(Vec<u8>, Vec<u8>)>> for DiaOracleKeyConvertor {
+	fn convert(spacwalk_oracle_key: OracleKey) -> Option<(Vec<u8>, Vec<u8>)> {
+		match spacwalk_oracle_key {
+			OracleKey::ExchangeRate(currency_id) => match currency_id {
+				CurrencyId::Token(token_symbol) => match token_symbol {
+					primitives::TokenSymbol::DOT =>
+						return Some((
+							DOT_DIA_BLOCKCHAIN.as_bytes().to_vec(),
+							DOT_DIA_SYMBOL.as_bytes().to_vec(),
+						)),
+					primitives::TokenSymbol::KSM =>
+						return Some((
+							KSM_DIA_BLOCKCHAIN.as_bytes().to_vec(),
+							KSM_DIA_SYMBOL.as_bytes().to_vec(),
+						)),
+					primitives::TokenSymbol::PEN => unimplemented!(),
+					primitives::TokenSymbol::AMPE => unimplemented!(),
+				},
+				CurrencyId::ForeignAsset(_) => unimplemented!(),
+				CurrencyId::Native => unimplemented!(),
+				CurrencyId::StellarNative => unimplemented!(),
+				CurrencyId::AlphaNum4 { .. } => unimplemented!(),
+				CurrencyId::AlphaNum12 { .. } => unimplemented!(),
+			},
+			OracleKey::FeeEstimation => Some((vec![6u8], vec![])),
+		}
+	}
+}
+
+impl Convert<(Vec<u8>, Vec<u8>), Option<OracleKey>> for DiaOracleKeyConvertor {
+	fn convert(dia_oracle_key: (Vec<u8>, Vec<u8>)) -> Option<OracleKey> {
+		let (blockchain, symbol) = dia_oracle_key;
+		let blockchain = String::from_utf8(blockchain);
+		let symbol = String::from_utf8(symbol);
+		match (blockchain, symbol) {
+			(Ok(blockchain), Ok(symbol)) => {
+				if blockchain == DOT_DIA_BLOCKCHAIN && symbol == DOT_DIA_SYMBOL {
+					return Some(OracleKey::ExchangeRate(CurrencyId::Token(
+						primitives::TokenSymbol::DOT,
+					)))
+				} else if blockchain == KSM_DIA_BLOCKCHAIN && symbol == KSM_DIA_SYMBOL {
+					return Some(OracleKey::ExchangeRate(CurrencyId::Token(
+						primitives::TokenSymbol::KSM,
+					)))
+				} else {
+					return None
+				}
+			},
+			(_, _) => return None,
+		}
+	}
+}
+
+pub struct DiaOracleAdapter<
+	DiaPallet: DiaOracle,
+	UnsignedFixedPoint,
+	Moment,
+	ConvertKey,
+	ConvertPrice,
+	ConvertMoment,
+>(
+	marker::PhantomData<(
+		DiaPallet,
+		UnsignedFixedPoint,
+		Moment,
+		ConvertKey,
+		ConvertPrice,
+		ConvertMoment,
+	)>,
+);
+
+impl<Dia, UnsignedFixedPoint, Moment, ConvertKey, ConvertPrice, ConvertMoment>
+	DataProviderExtended<OracleKey, TimestampedValue<UnsignedFixedPoint, Moment>>
+	for DiaOracleAdapter<Dia, UnsignedFixedPoint, Moment, ConvertKey, ConvertPrice, ConvertMoment>
+where
+	Dia: DiaOracle,
+	ConvertKey: Convert<OracleKey, Option<(Vec<u8>, Vec<u8>)>>
+		+ Convert<(Vec<u8>, Vec<u8>), Option<OracleKey>>,
+	ConvertPrice: Convert<u128, Option<UnsignedFixedPoint>>,
+	ConvertMoment: Convert<u64, Option<Moment>>,
+{
+	fn get_no_op(key: &OracleKey) -> Option<TimestampedValue<UnsignedFixedPoint, Moment>> {
+		let dia_key: Option<(Vec<u8>, Vec<u8>)> = ConvertKey::convert(key.clone());
+		let Some((blockchain,symbol)) = dia_key else {
```

Comment:
since this function already returns an `Option`, you can actually use the [`?` operator](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/reference/expressions/operator-expr.html#the-question-mark-operator);
```rust
let (blockchain,symbol) = ConvertKey::convert(key.clone())?;
```

Reply:
thanks! You are right!